client/cos.go

```diff
@@ -0,0 +1,181 @@
+package client
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/cosigner/msgs"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+	"github.com/sirupsen/logrus"
+)
+
+type CosignerProvider struct {
+	Issuer       string
+	CallbackPath string
+}
+
+func (p *CosignerProvider) GetIssuer() string {
+	return p.Issuer
+}
+
+func (c *CosignerProvider) RequestToken(ctx context.Context, signer crypto.Signer, pkt *pktoken.PKToken, redirCh chan string) (*pktoken.PKToken, error) {
+	// Find an unused port
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to bind to an available port: %w", err)
+	}
+
+	port := listener.Addr().(*net.TCPAddr).Port
+	host := fmt.Sprintf("localhost:%d", port)
+	redirectURI := fmt.Sprintf("http://%s%s", host, c.CallbackPath)
+
+	mux := http.NewServeMux()
+	server := &http.Server{
+		Addr:    host,
+		Handler: mux,
+	}
+
+	logrus.Infof("listening on http://%s/", host)
+	logrus.Info("press ctrl+c to stop")
+	go func() {
+		err := server.Serve(listener)
+		if err != nil && err != http.ErrServerClosed {
+			logrus.Error(err)
+		}
+	}()
+	defer server.Shutdown(ctx)
+
+	ch := make(chan []byte)
+	errCh := make(chan error)
```

Comment:
Let's see if I understand what you are saying:

`errCh := make(chan error)` is equivalent to `errCh := make(chan error, 0)`. This means if an error occursin the CallbackPath-handler we before we reach `case err := <-errCh:` then the CallbackPath-handler function be blocked. Then if `ctx.Done()` is called, we might never reach `case err := <-errCh:` and CallbackPath-handler will block indefinitely (or at least until process tear down). 

Where as if 
```golang
ch := make(chan []byte, 1)
errCh := make(chan error, 1)
```
CallbackPath-handler will not block and always execute the return statement following each `ch` or `errCh` send?

Reply:
Fixed to have 1 buffer channels

Reply:
> Let's see if I understand what you are saying:
> ...
> Then if ctx.Done() is called, we might never reach case err := <-errCh: and CallbackPath-handler will block indefinitely (or at least until process tear down).

> CallbackPath-handler will not block and always execute the return statement following each ch or errCh send?

Yes that was my original concern. The buffer of size `1` definitely helps alleviate the issue, but on second thought, I guess technically there is still a very small chance for that handler to still be blocking in case two errors are received before `server.Shutdown()` can be called after the `ctx` is cancelled. So the buffer size only makes the issue less likely. Maybe this isn't too much of a concern though--let's just keep it with buffer of size 1 for now.